include/irr/asset/CTexturePacker.h

```diff
@@ -163,83 +133,144 @@ class ITexturePacker
-        m_megaimg->setBufferAndRegions(std::move(buffer), regions);
+            m_pageTable->setBufferAndRegions(std::move(buf), regions);
+        }
     }
 
     STexOffset pack(const ICPUImage* _img, const ICPUImage::SSubresourceRange& _subres)
     {
         //assert(m_megaimg->getCreationParameters().format==_img->getCreationParameters().format);
 
-        const uint32_t texelSz = getTexelOrBlockBytesize(m_megaimg->getCreationParameters().format);
+        const uint32_t texelSz = getTexelOrBlockBytesize(m_physAddrTex->getCreationParameters().format);
         auto offsets = alloc(_img, _subres);
         for (const auto& reg : _img->getRegions())
         {
-            if (reg.imageSubresource.mipLevel>=_subres.baseMipLevel && reg.imageSubresource.mipLevel<(_subres.baseMipLevel+offsets->size()))
+            if (reg.imageSubresource.mipLevel<_subres.baseMipLevel || reg.imageSubresource.mipLevel>=(_subres.baseMipLevel+offsets->size()))
                 continue;
 
-            const uint32_t j = reg.imageSubresource.mipLevel-_subres.baseMipLevel;
+            const uint32_t j = reg.imageSubresource.mipLevel-_subres.baseMipLevel;//"relative LoD"
 
-            const core::vector2du32_SIMD texOffset = pageCoords((*offsets)[j].x(), (*offsets)[j].y(), m_pgSzxy);
-            const uint64_t bufOffset = static_cast<uint64_t>(m_megaimg->getCreationParameters().extent.width * m_megaimg->getCreationParameters().extent.height * (*offsets)[j].layer() + texOffset.y * m_megaimg->getCreationParameters().extent.width + texOffset.x)* texelSz;
+            const core::vector3du32_SIMD texOffset = pageCoords((*offsets)[j], m_pgSzxy);
+            const uint64_t bufOffset = static_cast<uint64_t>(m_physAddrTex->getCreationParameters().extent.width*m_physAddrTex->getCreationParameters().extent.height*texOffset.z + texOffset.y*m_physAddrTex->getCreationParameters().extent.width + texOffset.x) * texelSz;
 
             const uint8_t* src = reinterpret_cast<const uint8_t*>(_img->getBuffer()->getPointer())+reg.bufferOffset;
-            uint8_t* dst = reinterpret_cast<uint8_t*>(m_megaimg->getBuffer()->getPointer()) + bufOffset;
+            uint8_t* dst = reinterpret_cast<uint8_t*>(m_physAddrTex->getBuffer()->getPointer()) + bufOffset;
             const uint32_t pitch = reg.bufferRowLength*texelSz;
             for (uint32_t i = 0u; i < reg.imageExtent.height; ++i)
             {
-                memcpy(dst + ((reg.imageOffset.y + i)*m_megaimg->getCreationParameters().extent.width + reg.imageOffset.x)*texelSz, src+i*pitch, reg.imageExtent.width*texelSz);
+                memcpy(dst + ((reg.imageOffset.y + i)*m_physAddrTex->getCreationParameters().extent.width + reg.imageOffset.x)*texelSz, src+i*pitch, reg.imageExtent.width*texelSz);
             }
         }
+        //pack tail mips into 1 page (if needed)
+        if (offsets->back().hasMipTailAddr())
```

Comment:
yeah this is screwed, cause `STexOffset` is not supposed to store that

Reply:
https://github.com/buildaworldnet/IrrlichtBAW/pull/424#discussion_r392975190